### PR TITLE
[Paddle-ASP]Support sharding training for the Nvidia's ASP(2:4 sparsity) functionality

### PR DESCRIPTION
**Pull request by minghaoBD**
### PR types
Bug fixes 

### PR changes
Others

### Describe
Nvidia has implemented 2:4 sparsity code in PaddlePaddle, supporting fleet distributed training. But when we are trying to train with sharding strategy (the model parallel paradigm in PaddlePaddle), GPU:0 will always be OOM while other GPUs seems normal. 

After fix, developers should pass in an argument: shading=True when calling sparsity.prune_model() with the sharding strategy. Otherwise no difference when using the APIs.

### CODE REVIEW

**Review by qingqing01 — COMMENTED** [review]
Please add UT

File: python/paddle/fluid/contrib/sparsity/asp.py [hunk 1]
```diff
@@ -150,7 +155,8 @@ def prune_model(main_program=None,
                 n=2,
                 m=4,
                 mask_algo='mask_1d',
-                with_mask=True):
+                with_mask=True,
+                sharding=False):
```
Comment (wanghaoshuang) [inline python/paddle/fluid/contrib/sparsity/asp.py:159]: 可以让用户直接传一个place么？那种方式更好理解？

Comment (minghaoBD) [reply python/paddle/fluid/contrib/sparsity/asp.py:159]: 理解是好理解，但是我们需要给用户额外说明。
此外，即使说明了，也有经常place设置错误的风险，然后出bug。
我的理解是，place在prune_model里面，代码稳定性更高一些呢？

**Comment by minghaoBD** [comment]
> Please add UT

added tests for optimizer compatibility and modified prune_model API.

**Comment by minghaoBD** [comment]
> Please add UT

added tests for optimizer compatibility and modified prune_model API.

**Review by wanghaoshuang — APPROVED** [review]
LGTM.

**Comment by wanghaoshuang** [comment]
请修改下PR标题，方便后续检索和管理自己的工作。

**Comment by minghaoBD** [comment]
> 请修改下PR标题，方便后续检索和管理自己的工作。

Done, thanks

**Review by JZ-LIANG — APPROVED** [review]
LGTM for sharding

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Review by TCChenlong — APPROVED** [review]
LGTM

**Pull request merged**